shared/src/event_handling.rs

```diff
@@ -0,0 +1,181 @@
+// transaction.
+const INSERT_EVENT_BATCH_SIZE: usize = 10_000;
+
+pub struct EventHandler<C: EventRetrieving, S: EventStoring<C::Event>> {
+    contract: C,
+    store: S,
+    last_handled_block: Option<u64>,
+}
+
+/// `EventStoring` is used by `EventHandler` for the purpose of giving the user freedom
+/// in how, where and which events are stored.
+///
+/// # Examples
+/// Databases: might transform, filter and classify which events are inserted
+/// HashSet: For less persistent (in memory) storing, insert events into a set.
+#[async_trait::async_trait]
+pub trait EventStoring<T> {
+    /// Returns ok, on successful execution, otherwise an appropriate error
+    ///
+    /// # Arguments
+    /// * `events` the contract events to be saved by the implementer
+    /// * `range` optionally indicates a particular range of blocks on which to operate.
+    async fn save_events(
+        &self,
+        events: Vec<EthcontractEvent<T>>,
+        range: Option<RangeInclusive<BlockNumber>>,
+    ) -> Result<()>;
+
+    async fn last_event_block(&self) -> Result<u64>;
+}
+
+pub trait EventRetrieving {
+    type Event: ParseLog;
+    fn get_events(&self) -> AllEventsBuilder<DynTransport, Self::Event>;
+    fn web3(&self) -> Web3<DynTransport>;
+}
+
+impl<C, S> EventHandler<C, S>
+where
+    C: EventRetrieving,
+    S: EventStoring<C::Event>,
+{
+    pub fn new(contract: C, store: S, start_sync_at_block: Option<u64>) -> Self {
+        Self {
+            contract,
+            store,
+            last_handled_block: start_sync_at_block,
+        }
+    }
+
+    async fn event_block_range(&self) -> Result<RangeInclusive<BlockNumber>> {
+        let web3 = self.contract.web3();
+        // Instead of using only the most recent event block from the db we also store the last
+        // handled block in self so that during long times of no events we do not query needlessly
+        // large block ranges.
+        let last_handled_block = match self.last_handled_block {
+            Some(block) => block,
+            None => self.store.last_event_block().await?,
+        };
+        let current_block = web3
+            .eth()
+            .block_number()
+            .await
+            .context("failed to get current block")?
+            .as_u64();
+        let from_block = last_handled_block.saturating_sub(MAX_REORG_BLOCK_COUNT);
+        anyhow::ensure!(
+            from_block <= current_block,
+            format!(
+                "current block number according to node is {} which is more than {} blocks in the \
+                 past compared to last handled block {}",
+                current_block, MAX_REORG_BLOCK_COUNT, last_handled_block
+            )
+        );
+        Ok(BlockNumber::Specific(from_block)..=BlockNumber::Latest(current_block))
+    }
+
+    /// Get new events from the contract and insert them into the database.
+    pub async fn update_events(&mut self) -> Result<()> {
```

Comment:
Maybe the logic on when to trigger this could also become shared functionality (like an UpdatingEventHandler). It probably makes sense to update events whenever we notice a new block. We already have `current_block_stream` which tells us whenever there is a new block and which I believe currently powers the run_maintenance logic.

Reply:
This is a cool idea. We could combine a `current_block_stream` with an `EventHandler` to make something that "just works". I would leave that for a follow up PR though.

Reply:
Created an issue #613 for this point.